From c01f049d638a48e2ac01ded9baaca3fb8589107c Mon Sep 17 00:00:00 2001
From: majing921201 <jing1.ma@intel.com>
Date: Fri, 11 Feb 2022 10:16:42 +0800
Subject: [PATCH 01/12] add SparseXPU to dispatch key set (#25)

Signed-off-by: Ma, Jing1 <jing1.ma@intel.com>
---
 c10/core/DispatchKeySet.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/c10/core/DispatchKeySet.h b/c10/core/DispatchKeySet.h
index b1f5f04524..2ec0984705 100644
--- a/c10/core/DispatchKeySet.h
+++ b/c10/core/DispatchKeySet.h
@@ -259,6 +259,7 @@ constexpr DispatchKeySet autogradother_backends = DispatchKeySet(
      DispatchKey::SparseCUDA,
      DispatchKey::SparseHIP,
      DispatchKey::SparseVE,
+     DispatchKey::SparseXPU,
      DispatchKey::SparseCsrCPU,
      DispatchKey::SparseCsrCUDA,
      DispatchKey::Meta});
-- 
2.25.1

