From a5425fa5e4491d51d4b636173262e998ee21544f Mon Sep 17 00:00:00 2001
From: xiaolil1 <xiaoli.liu@intel.com>
Date: Tue, 15 Feb 2022 14:01:53 +0800
Subject: [PATCH 02/29] Add QuantizedXPU to DispatchKeySet (#28)

---
 c10/core/DispatchKeySet.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/c10/core/DispatchKeySet.h b/c10/core/DispatchKeySet.h
index 2ec0984705..2ad7a0fb05 100644
--- a/c10/core/DispatchKeySet.h
+++ b/c10/core/DispatchKeySet.h
@@ -253,6 +253,7 @@ constexpr DispatchKeySet autogradother_backends = DispatchKeySet(
      DispatchKey::Metal,
      DispatchKey::QuantizedCPU,
      DispatchKey::QuantizedCUDA,
+     DispatchKey::QuantizedXPU,
      DispatchKey::CustomRNGKeyId,
      DispatchKey::MkldnnCPU,
      DispatchKey::SparseCPU,
-- 
2.25.1

