From 76929b8e0ff49a6bf86bd7b2dfb4099293acceed Mon Sep 17 00:00:00 2001
From: "Huaiyu, Zheng" <huaiyu.zheng@intel.com>
Date: Thu, 10 Oct 2024 19:36:36 +0800
Subject: [PATCH 3/6] Update torch-xpu-ops pin to 02060c6 (#303)

---
 third_party/xpu.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/third_party/xpu.txt b/third_party/xpu.txt
index 69606f14a7a..529b1942193 100644
--- a/third_party/xpu.txt
+++ b/third_party/xpu.txt
@@ -1 +1 @@
-7e3d00acea9f0d3728048a5b2743de20d55c64ba
+02060c616661034286adb9d5ba493e234b8401c3
-- 
2.34.1

