From 7eaf44e4ced6303b8e06cd412eb3787d07698b1c Mon Sep 17 00:00:00 2001
From: "Yu, Guangye" <guangye.yu@intel.com>
Date: Mon, 22 May 2023 09:57:58 +0000
Subject: [PATCH 05/25] add xpu device type map to dispatch key

---
 torch/library.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/torch/library.h b/torch/library.h
index 71b190e8a3e..a11fe6165d3 100644
--- a/torch/library.h
+++ b/torch/library.h
@@ -362,6 +362,8 @@ inline CppFunction dispatch(c10::DeviceType type, Func&& raw_f) {
         return c10::DispatchKey::XLA;
       case c10::DeviceType::Lazy:
         return c10::DispatchKey::Lazy;
+      case c10::DeviceType::XPU:
+        return c10::DispatchKey::XPU;
       case c10::DeviceType::MPS:
         return c10::DispatchKey::MPS;
       case c10::DeviceType::Meta:
-- 
2.34.1

