From 414928de5c0e00ea4d21fbef5297ddd90c887cd7 Mon Sep 17 00:00:00 2001
From: "Yu, Guangye" <106960996+guangyey@users.noreply.github.com>
Date: Tue, 19 Sep 2023 09:39:36 +0800
Subject: [PATCH 05/17] fix hpu deserialization bug (#165)

---
 torch/serialization.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/torch/serialization.py b/torch/serialization.py
index b1399eebbf9..4c7301ab4bb 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -300,9 +300,9 @@ def validate_hpu_device(location):
 
 
 def _hpu_deserialize(obj, location):
-    hpu = getattr(torch, "hpu", None)
-    assert hpu is not None, "HPU device module is not loaded"
     if location.startswith('hpu'):
+        hpu = getattr(torch, "hpu", None)
+        assert hpu is not None, "HPU device module is not loaded"
         device = validate_hpu_device(location)
         if getattr(obj, "_torch_load_uninitialized", False):
             with hpu.device(device):
-- 
2.34.1

