From a5dec8aa8b6bb33e8a25ceeffa3e0fc00b62f346 Mon Sep 17 00:00:00 2001
From: "Yu, Guangye" <106960996+guangyey@users.noreply.github.com>
Date: Fri, 21 Apr 2023 12:05:18 +0800
Subject: [PATCH 05/10] add xpu device type map to dispatch key (#105) (#120)

---
 torch/library.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/torch/library.h b/torch/library.h
index 71b190e8a3e..a11fe6165d3 100644
--- a/torch/library.h
+++ b/torch/library.h
@@ -362,6 +362,8 @@ inline CppFunction dispatch(c10::DeviceType type, Func&& raw_f) {
         return c10::DispatchKey::XLA;
       case c10::DeviceType::Lazy:
         return c10::DispatchKey::Lazy;
+      case c10::DeviceType::XPU:
+        return c10::DispatchKey::XPU;
       case c10::DeviceType::MPS:
         return c10::DispatchKey::MPS;
       case c10::DeviceType::Meta:
-- 
2.25.1

