From adbcccaa2248aa881ff5282ddf34e4eede6cb25d Mon Sep 17 00:00:00 2001
From: min-jean-cho <min.jean.cho@intel.com>
Date: Fri, 22 Sep 2023 03:09:51 -0700
Subject: [PATCH 07/30] add path for DPC++ SYCL device code (#167)

---
 c10/util/Float8_e4m3fn.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/c10/util/Float8_e4m3fn.h b/c10/util/Float8_e4m3fn.h
index 11b423cc517..8d3e339ca61 100644
--- a/c10/util/Float8_e4m3fn.h
+++ b/c10/util/Float8_e4m3fn.h
@@ -98,6 +98,10 @@ inline C10_HOST_DEVICE float fp8e4m3fn_to_fp32_value(uint8_t input) {
    */
 #if defined(__CUDA_ARCH__)
   uint32_t renorm_shift = __clz(nonsign);
+#elif defined(__SYCL_DEVICE_ONLY__)
+  // Note: zero is not a supported input into `__builtin_clz`
+  uint32_t renorm_shift =
+      nonsign != 0 ? __builtin_clz(nonsign) : sizeof(uint32_t) * CHAR_BIT;
 #elif defined(_MSC_VER)
   unsigned long nonsign_bsr;
   _BitScanReverse(&nonsign_bsr, (unsigned long)nonsign);
-- 
2.34.1

