From 85d5016e2ea7922b6a72fd9ca23db7017bfc7685 Mon Sep 17 00:00:00 2001
From: chengjunlu <chengjun.lu@intel.com>
Date: Tue, 6 Jul 2021 16:17:11 +0800
Subject: [PATCH 12/17] 12.Add thread local guard in backward hooks (#10)

---
 torch/csrc/autograd/engine.cpp | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/torch/csrc/autograd/engine.cpp b/torch/csrc/autograd/engine.cpp
index 62ca26e469..d9a72ea02c 100644
--- a/torch/csrc/autograd/engine.cpp
+++ b/torch/csrc/autograd/engine.cpp
@@ -493,10 +493,13 @@ void GraphTask::exec_post_processing() {
   // WARNING: Don't use a range-for loop here because more callbacks may be
   // added in between callback calls, so iterators may become invalidated.
   // NOLINTNEXTLINE(modernize-loop-convert)
-  for (size_t i = 0; i < final_callbacks_.size(); ++i) {
-    cb_lock.unlock();
-    final_callbacks_[i]();
-    cb_lock.lock();
+  {
+    at::ThreadLocalStateGuard guard(this->thread_locals_);
+    for (size_t i = 0; i < final_callbacks_.size(); ++i) {
+      cb_lock.unlock();
+      final_callbacks_[i]();
+      cb_lock.lock();
+    }
   }
 
   // Syncs leaf streams with default streams (if necessary)
@@ -686,6 +689,7 @@ static variable_list call_function(
 
   if(has_post_hooks){
     // NOLINTNEXTLINE(bugprone-use-after-move)
+    at::ThreadLocalStateGuard guard(graph_task->thread_locals_);
     return call_post_hooks(fn, std::move(outputs), inputs);
   }
   return outputs;
-- 
2.25.1

