From 4a3d0d4e39ad1017a4ff688e199ac90ff8de2b43 Mon Sep 17 00:00:00 2001
From: "Yu, Guangye" <106960996+guangyey@users.noreply.github.com>
Date: Sat, 17 Jun 2023 00:58:30 +0800
Subject: [PATCH 14/25] Align to masked_fill (#134)

---
 torch/_refs/__init__.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/torch/_refs/__init__.py b/torch/_refs/__init__.py
index f5e6bd70f3e..9c0f1f22e12 100644
--- a/torch/_refs/__init__.py
+++ b/torch/_refs/__init__.py
@@ -4885,8 +4885,8 @@ def masked_fill(a: TensorLikeType, mask: TensorLikeType, value: TensorOrNumberLi
             value_ndim == 0,
             lambda: f"only supports a 0-dimensional value tensor, but got tensor with {value_ndim} dimension",
         )
-        # `masked_fill` allows cpu scalar to be moved to cuda but not otherwise.
-        is_cpu_scalar = a.device.type == "cuda" and value.device.type == "cpu"
+        # `masked_fill` allows cpu scalar to be moved to cuda and xpu but not otherwise.
+        is_cpu_scalar = a.device.type in ["cuda", "xpu"] and value.device.type == "cpu"
         check(
             is_cpu_scalar or value.device == a.device,
             lambda: "Expected `value` to be on same device as `a`",
-- 
2.34.1

