From 5f22c2c503d7e424f2d2953fec333e0e8e901d4d Mon Sep 17 00:00:00 2001
From: "Yu, Guangye" <106960996+guangyey@users.noreply.github.com>
Date: Wed, 29 Nov 2023 18:08:57 +0800
Subject: [PATCH 20/26] fix deterministic issue (#199)

---
 benchmarks/dynamo/common.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/benchmarks/dynamo/common.py b/benchmarks/dynamo/common.py
index 9539e45c2de..8a8cb86331d 100644
--- a/benchmarks/dynamo/common.py
+++ b/benchmarks/dynamo/common.py
@@ -380,6 +380,8 @@ def patch_torch_manual_seed():
 
         if not torch.cuda._is_in_bad_fork():
             torch.cuda.manual_seed_all(seed)
+        if not torch.xpu._is_in_bad_fork():
+            torch.xpu.manual_seed_all(seed)
         return default_generator.manual_seed(seed)
 
     torch.manual_seed = deterministic_torch_manual_seed
-- 
2.34.1

