From 28c69d4b4a71a71b2919cc3cb4ad22d8edd933ea Mon Sep 17 00:00:00 2001
From: wuhuikx <hui.h.wu@intel.com>
Date: Fri, 5 Aug 2022 15:00:23 +0800
Subject: [PATCH 24/35] Speedup Philox on XPU device (#58)

---
 aten/src/ATen/core/PhiloxRNGEngine.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/aten/src/ATen/core/PhiloxRNGEngine.h b/aten/src/ATen/core/PhiloxRNGEngine.h
index d075d7dd6f..6097f212cf 100644
--- a/aten/src/ATen/core/PhiloxRNGEngine.h
+++ b/aten/src/ATen/core/PhiloxRNGEngine.h
@@ -12,6 +12,10 @@
 #include <cuda.h>
 #endif
 
+#ifdef SYCL_LANGUAGE_VERSION
+#include <CL/sycl.hpp>
+#endif
+
 #include <ATen/core/Array.h>
 #include <c10/macros/Macros.h>
 #include <c10/util/Exception.h>
@@ -173,6 +177,9 @@ private:
     #ifdef __CUDA_ARCH__
       *result_high = __umulhi(a, b);
       return a*b;
+    #elif __SYCL_DEVICE_ONLY__
+      *result_high = sycl::mul_hi(a, b);
+      return a*b;
     #else
       const uint64_t product = static_cast<uint64_t>(a) * b;
       *result_high = static_cast<uint32_t>(product >> 32);
-- 
2.25.1

