From f24e7798951208a4409cef909ff9615065577329 Mon Sep 17 00:00:00 2001
From: Lei Zhenyuan <zhenyuan.lei@intel.com>
Date: Fri, 1 Mar 2024 12:33:22 +0800
Subject: [PATCH 26/26] expose nested header to wheel (#206)

---
 setup.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/setup.py b/setup.py
index 17bf16b89a7..b58603edd6d 100644
--- a/setup.py
+++ b/setup.py
@@ -1236,6 +1236,7 @@ def main():
         "include/ATen/native/hip/*.h",
         "include/ATen/native/hip/*.cuh",
         "include/ATen/native/mps/*.h",
+        "include/ATen/native/nested/*.h",
         "include/ATen/native/quantized/*.h",
         "include/ATen/native/quantized/cpu/*.h",
         "include/ATen/native/utils/*.h",
-- 
2.34.1

