From 7897dbe6b8868cfe318e52bab1e233acc07b40f3 Mon Sep 17 00:00:00 2001
From: "Cui, Yifeng" <yifeng.cui@intel.com>
Date: Thu, 7 Mar 2024 13:30:38 +0800
Subject: [PATCH 27/31] Hide MKL symbols (#211)

---
 cmake/public/mkl.cmake | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/cmake/public/mkl.cmake b/cmake/public/mkl.cmake
index 2701c0700f7..d285a92600f 100644
--- a/cmake/public/mkl.cmake
+++ b/cmake/public/mkl.cmake
@@ -15,3 +15,15 @@ set_property(
 set_property(
   TARGET caffe2::mkl PROPERTY INTERFACE_LINK_DIRECTORIES
   ${MKL_ROOT}/lib ${MKL_ROOT}/lib/intel64 ${MKL_ROOT}/lib/intel64_win ${MKL_ROOT}/lib/win-x64)
+
+if(${USE_STATIC_MKL})
+  foreach(MKL_LIB_PATH IN LISTS ONEMKL_LIBRARIES)
+    if(EXISTS "${MKL_LIB_PATH}")
+      get_filename_component(MKL_LIB_NAME "${MKL_LIB_PATH}" NAME)
+      # Match archive libraries starting with "libmkl_"
+      if(MKL_LIB_NAME MATCHES "^libmkl_" AND MKL_LIB_NAME MATCHES ".a$")
+        target_link_options(caffe2::mkl INTERFACE "-Wl,--exclude-libs,${MKL_LIB_NAME}")
+      endif()
+    endif()
+  endforeach()
+endif()
-- 
2.34.1

