From 6f57f433f1cd2c10f82fd0ebf41c1a1762e13424 Mon Sep 17 00:00:00 2001
From: Jing Xu <jing.xu@intel.com>
Date: Tue, 18 Jun 2024 20:03:25 +0900
Subject: [PATCH 28/32] fix numpy2 upgrade issue (#258)

---
 torch/csrc/utils/tensor_numpy.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/torch/csrc/utils/tensor_numpy.cpp b/torch/csrc/utils/tensor_numpy.cpp
index c18dec2bbdd..de90dfe04a7 100644
--- a/torch/csrc/utils/tensor_numpy.cpp
+++ b/torch/csrc/utils/tensor_numpy.cpp
@@ -414,7 +414,11 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
     TORCH_CHECK_VALUE(
         PyArray_DescrConverter(py_typestr, &descr), "cannot parse `typestr`");
     dtype = numpy_dtype_to_aten(descr->type_num);
+#if NPY_ABI_VERSION >= 0x02000000
+    dtype_size_in_bytes = PyDataType_ELSIZE(descr);
+#else
     dtype_size_in_bytes = descr->elsize;
+#endif
     TORCH_INTERNAL_ASSERT(dtype_size_in_bytes > 0);
   }
 
-- 
2.34.1

