From c61d29a68eaa4af9a5941e102a8ba59f989cf0d6 Mon Sep 17 00:00:00 2001
From: Zhiwei <zhiwei.yan@intel.com>
Date: Fri, 15 Mar 2024 15:18:03 +0800
Subject: [PATCH 29/30] Move observer to device (#203)

---
 torch/ao/quantization/observer.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/torch/ao/quantization/observer.py b/torch/ao/quantization/observer.py
index f5d24c45787..f8988f68c96 100644
--- a/torch/ao/quantization/observer.py
+++ b/torch/ao/quantization/observer.py
@@ -683,6 +683,8 @@ class PerChannelMinMaxObserver(UniformQuantizationObserverBase):
         if x_orig.numel() == 0:
             return x_orig
         x = x_orig.detach()  # avoid keeping autograd tape
+        self.min_val = self.min_val.to(x_orig.device)
+        self.max_val = self.max_val.to(x_orig.device)
         min_val = self.min_val
         max_val = self.max_val
         x_dim = x.size()
-- 
2.34.1

