diff --git a/cmake/Modules/FindMKL.cmake b/cmake/Modules/FindMKL.cmake
index 52aa9908a55..ecb59dafa31 100644
--- a/cmake/Modules/FindMKL.cmake
+++ b/cmake/Modules/FindMKL.cmake
@@ -263,10 +263,13 @@ MACRO(CHECK_ALL_LIBRARIES LIBRARIES OPENMP_TYPE OPENMP_LIBRARY _name _list _flag
       ELSE()
         IF(MSVC)
           SET(lib_names ${_library}_dll)
+          SET(lib_names_static ${_library})
+          # Both seek shared and static mkl library.
+          FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${lib_names} ${lib_names_static})
         ELSE()
           SET(lib_names ${_library})
+          FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${lib_names})
         ENDIF()
-        FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${lib_names})
       ENDIF()
       MARK_AS_ADVANCED(${_prefix}_${_library}_LIBRARY)
       IF(NOT (${_library} STREQUAL "tbb"))
diff --git a/setup.py b/setup.py
index 0383f0c80a4..aec336158b7 100644
--- a/setup.py
+++ b/setup.py
@@ -1095,7 +1095,6 @@ def main():
         "networkx",
         "jinja2",
         "fsspec",
-        'mkl>=2021.1.1,<=2021.4.0; platform_system == "Windows"',
     ]
 
     # Parse the command line and check the arguments before we proceed with
